kernel/src/net/socket/netlink/table.rs

```diff
@@ -0,0 +1,218 @@
+}
+
+impl NetlinkSocketTable {
+    pub const fn new() -> Self {
+        Self {
+            protocols: RwMutex::new(BTreeMap::new()),
+        }
+    }
+
+    /// Adds a new netlink protocol.
+    fn add_new_protocol(&self, protocol: NetlinkProtocolId) {
+        let mut protocols = self.protocols.write();
+        if protocols.contains_key(&protocol) {
+            return;
+        }
+        let new_protocol = RwMutex::new(ProtocolSocketTable::new(protocol));
+        protocols.insert(protocol, new_protocol);
+    }
+
+    pub fn bind(
+        &self,
+        protocol: NetlinkProtocolId,
+        addr: &NetlinkSocketAddr,
+    ) -> Result<BoundHandle> {
+        let protocols = self.protocols.read();
+
+        let Some(protocol_sockets) = protocols.get(&protocol) else {
+            return_errno_with_message!(Errno::EINVAL, "the netlink protocol does not exist")
+        };
+
+        let mut protocol_sockets = protocol_sockets.write();
+        protocol_sockets.bind(addr)
+    }
+}
+
+/// Bound socket table of a single netlink protocol.
+///
+/// Each table can have bound sockets for unicast
+/// and at most 32 groups for multicast.
+struct ProtocolSocketTable {
+    id: NetlinkProtocolId,
+    unicast_sockets: BTreeSet<PortNum>,
+    multicast_groups: Box<[MulticastGroup]>,
+}
+
+impl ProtocolSocketTable {
+    /// Creates a new table.
+    fn new(id: NetlinkProtocolId) -> Self {
+        let multicast_groups = (0u32..MAX_GROUPS).map(|_| MulticastGroup::new()).collect();
+        Self {
+            id,
+            unicast_sockets: BTreeSet::new(),
+            multicast_groups,
+        }
+    }
+
+    /// Binds a socket to the table.
+    /// Returns the bound handle.
+    ///
+    /// The socket will be bound to a port specified by `addr.port()`.
+    /// If `addr.port()` is zero, the kernel will assign a port,
+    /// typically corresponding to the process ID of the current process.
+    /// If the assigned port is already in use,
+    /// this function will try to allocate a random unused port.
+    ///
+    /// Additionally, this socket can join one or more multicast groups,
+    /// as specified in `addr.groups()`.
+    fn bind(&mut self, addr: &NetlinkSocketAddr) -> Result<BoundHandle> {
+        let port = if addr.port() != UNSPECIFIED_PORT {
+            addr.port()
+        } else {
+            let mut random_port = current!().pid();
+            while random_port == UNSPECIFIED_PORT || self.unicast_sockets.contains(&random_port) {
+                getrandom(random_port.as_bytes_mut()).unwrap();
+            }
+            random_port
+        };
+
+        if self.unicast_sockets.contains(&port) {
+            return_errno_with_message!(
+                Errno::EADDRINUSE,
+                "attempt to bind to an already-used port"
+            );
+        }
+
+        self.unicast_sockets.insert(port);
+
+        for group_id in addr.groups().ids_iter() {
+            debug_assert!(group_id < MAX_GROUPS);
```

Comment:
This is not necessary as we'll do the bound checking when accessing `multicast_groups` anyway.